### DISCUSSION TITLE
Automatic migrations from models similar to Django

### DISCUSSION THREAD

**Discussion by dhruvja** [diesel-rs/diesel#3669] [Ideas]
**Description**
I think having migrations be generated automatic based on the schema in the models file would be very useful. A migrations folder can be created and whenever something changes in the models file - For eg: if a column is added or removed, the change can be compared with the previous migration and it can be recorded. We can have a terminal command to generate migrations. 

**Inspiration**
This is inspired from Django where we have a models file where we can define the schema. It is written in python and it is pretty simple and extensible. Whenever the migrate command is run, the ORM records the changes done in models file and converts them into SQL statements to execute. This allows the user to have complete flexibility over the schema. The ORM detects everything which makes it easier. Even if the migrations folder is deleted, the migrations can be recovered from the current schema

**Proposal**
We can build this is a cargo package which has the automatic migrations compared to the current implementation where the migrations have to be written in SQL and then the models are generated.

I would prefer the other way. Let me know if you guys feel the same and if so we can build another package on top of diesel which has this automatic migrations feature. If it works well, then diesel can add this as a feature in their cargo package.

**Comment by weiznich** [comment]
Can you explain how the proposed feature would be different from the existing `diesel migration generate --diff-schema` command introduced in #3496?

**Reply by dhruvja** [reply]
So the diff-schema can generate sql commands. That's really great. I didn't know since I was following the getting started tutorial. Can we have a tutorial just that we include the diff schema in that tutorial? 


can we have a schema.rs and then run diff schema?

**Reply by weiznich** [reply]
The feature is currently quite new, so there is not much documentation about it yet (at least beside of what is written in the [release announcements](https://diesel.rs/news/2_1_0_release.html)). I don't want to add more and more stuff to the getting started guide as it already contains a huge amount of information. That written I would be happy if you have suggestions where to document this feature.

**Reply by dhruvja** [reply]
> The feature is currently quite new, so there is not much documentation about it yet (at least beside of what is written in the [release announcements](https://diesel.rs/news/2_1_0_release.html)). I don't want to add more and more stuff to the getting started guide as it already contains a huge amount of information. That written I would be happy if you have suggestions where to document this feature. 

I get it. Well I think we can have a note in getting started page saying that there is a feature to generate automatic migrations but it will be quite brief. We can then redirect to another page which can have some more information.

What do you think about this?

**Reply by weiznich** [reply]
I would accept a PR for that.

**Reply by dhruvja** [reply]
Thats lovely, thank you.

**Reply by dhruvja** [reply]
@weiznich  Regarding the getting started guides, it is managed in this repo right? https://github.com/sgrif/diesel.rs-website .

Curious since its not part of diesel organization.

**Reply by weiznich** [reply]
Yes that's the right repository. It's that way because of historic reason. Sage started diesel some years back and still manages the diesel.rs domain.
